src/crimson/os/seastore/transaction_manager.cc

```diff
@@ -63,8 +65,16 @@ TransactionManager::mount_ertr::future<> TransactionManager::mount()
   LOG_PREFIX(TransactionManager::mount);
   cache->init();
   segment_cleaner->mount(segment_manager);
-  return journal->replay([this](auto seq, auto paddr, const auto &e) {
-    return cache->replay_delta(seq, paddr, e);
+  return segment_cleaner->init_segments().safe_then(
+    [this](auto&& segments) {
+    return journal->replay(
+      std::move(segments),
+      [this](auto seq, auto paddr, const auto &e) {
+      auto fut = cache->replay_delta(seq, paddr, e);
+      segment_cleaner->update_journal_tail_committed(
+	cache->get_oldest_dirty_from().value_or(seq));
```

Comment:
I think update_journal_tail_target would probably work -- journal->open_for_write would then record the new journal_tail and update journal_tail_committed.

Reply:
Two problems.

First, I'd like to adhere to the existing pattern where TransactionManager updates journal_tail_target and journal_tail_committed only gets updated after journal_tail is committed.

Second, updating journal_tail_committed before journal_tail is actually committed would allow the SegmentCleaner to reclaim and reset segments that are after the most recent journal_tail value committed to disk.  It wouldn't actually destroy any deltas we need, but it would violate a useful invariant and break replay as written.

Reply:
Got it:-)

Reply:
@athanatos Hi, sam. I'm trying to correct this part of the code and run into a problem. I changed to use `SegmentCleaner::update_journal_tail_target()` here, however, the assert `assert(journal_tail_target == journal_seq_t() || target >= journal_tail_target);` is triggered. It seems that this is because the RootBlock who has a null `dirty_from_or_retired_at` is added to the cache's dirty list, and `Cache::get_oldest_dirty_from()` always return empty until the first root delta is found.

I checked the commit message of the RootBlock init code and it says: "ROOT blocks never show up in the transaction fresh or mutated lists nor do they show up in the cache extent set or dirty list". It seems that the code is not consistent with the commit message. Should I modify it to prevent the root block from being added to the cache's dirty list? Thanks:-)

Reply:
I found the follow up commit that allow root block in the dirty list, but I think it doesn't need to be dirty at the start, am I right? I think the root block can be dirty once it's modified or a root delta is found.

Reply:
No, the dirty_from value for the root should always be the location of its most recent delta.  Presumably, the replay code just needs to initialize that correctly for root when it finds it.

Reply:
Yes, I just meant that we can set the root to be clean until we find the first root delta, this won't affect its dirty_from value, right? Only its state is affected.

Reply:
Right now, the problem is root has a null dirty_from value and is set to DIRTY at the beginning, which makes the `Cache:: get_oldest_dirty_from()` method returns that null dirty_from even if there are valid dirty_froms in the dirty list. At the time of replay, before a root delta is met there can be other deltas, the above facts would lead to the transaction manager always updating the journal_tail_target with the most recent journal seq not the oldest dirty_from.

Reply:
> No, the dirty_from value for the root should always be the location of its most recent delta as that would be the delta we need to replay to recover the root. Presumably, the replay code just needs to initialize that correctly for root when it finds it.

@athanatos Hi, sam, I think there must be some misunderstanding here. I rechecked the commit [8bd35ed](https://github.com/ceph/ceph/pull/41803/commits/8bd35ed8d2a6dc3886d518f03f513448d5c5a298), and it looks to me that it doesn't conflict with what you said.

The issue is as follows:

1. the root block is set to dirty and added to the cache's dirty list;
2. the journal replays deltas, the first several of which are non-root, and tries to `update_journal_tail_target` with `cache->get_oldest_dirty_from().value_or(seq)`;
3. as the root block whose dirty_from is null at this moment(because the journal hasn't found root deltas yet), and is in the cache's dirty list, the "Cache::get_oldest_dirty_from()" would return an empty std::optional<journal_seq_t>, which would make `cache->get_oldest_dirty_from().value_or(seq)` return the "seq" which is the current journal seq;
4. as a result, the `SegmentCleaner::journal_tail_target` would go beyond the oldest dirty_from of Cache;
5. when a root delta is found and the root block is updated to hold a valid dirty_from, the next `Cache::get_oldeset_dirty_from()` would return the actual oldest dirty_from in Cache which is already less than `SegmentCleaner::journal_tail_target`, so when the next `SegmentCleaner::update_journal_tail_target()` call is triggered, the assert `journal_tail_target == journal_seq_t() || target >= journal_tail_target` in `SegmentCleaner::update_journal_tail_target()` would fail.

To fix this issue, I chose to set the root block's state to CLEAN (maybe INVALID should be more appropriate), so it wouldn't go into the cache's dirty list before the first root delta is met by the journal (or it is modified in the case of mkfs). WIth this approach, the above case won't happen. And once the first root delta is met or it is modified, it will go into the dirty list and all processing after that is the same as before.

I think this won't cause the recovery of the root block to be wrong, am I right? Thanks:-)

Sorry that I didn't elaborate the issue at the beginning of this discussion:-)

Reply:
Ah, I see.  That seems reasonable.  Perhaps assert root is dirty and in the dirty list at the end of replay?

Reply:
> Ah, I see. That seems reasonable. Perhaps assert root is dirty and in the dirty list at the end of replay?

Yes, will do:-)

Reply:
done:-)